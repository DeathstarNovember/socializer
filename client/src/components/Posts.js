import React, { useCallback } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import produce from "immer";
import { Feed, QueryResult } from "components";

export const GET_POSTS = gql`
  query GetPosts {
    posts {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

export const POSTS_SUBSCRIPTION = gql`
  subscription onPostCreated {
    postCreated {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

const Posts = () => {
  const { subscribeToMore, ...queryResult } = useQuery(GET_POSTS);
  const subscribeToNew = useCallback(
    () =>
      subscribeToMore({
        document: POSTS_SUBSCRIPTION,
        updateQuery: (prev, { subscriptionData }) => {
          if (!subscriptionData.data) return prev;
          const newPost = subscriptionData.data.postCreated;

          // Check that we don't already have the
          // post stored.
          if (prev.posts.find((post) => post.id === newPost.id)) {
            return prev;
          }

          return produce(prev, (next) => {
            next.posts.unshift(newPost);
          });
        },
      }),
    [],
  );

  return (
    <>
      <h4>Feed</h4>
      <QueryResult {...queryResult}>
        {({ data }) => (
          <Feed
            feedType="post"
            items={data.posts}
            subscribeToNew={subscribeToNew}
          />
        )}
      </QueryResult>
    </>
  );
};

export default Posts;
