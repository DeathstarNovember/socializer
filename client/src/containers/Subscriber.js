import { useEffect } from "react";

const Subscriber = ({ subscribeToNew, children }) => {
  useEffect(() => {
    subscribeToNew();
  }, [subscribeToNew]);

  return children;
};

export default Subscriber;
