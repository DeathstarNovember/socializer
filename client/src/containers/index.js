export { default as StateProvider } from "./StateProvider";
export { default as Subscriber } from "./Subscriber";
