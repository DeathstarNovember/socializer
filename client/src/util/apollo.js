import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
