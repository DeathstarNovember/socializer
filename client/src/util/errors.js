import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
