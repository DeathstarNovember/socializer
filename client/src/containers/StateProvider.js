import React, { useEffect, useState } from "react";
import { useApolloClient } from "@apollo/react-hooks";
import Cookies from "js-cookie";
import { refreshSocket } from "util/apollo";
import { AuthContext, ChatContext } from "util/context";

const StateProvider = ({ initialToken, initialUserId, socket, children }) => {
  const client = useApolloClient();
  const [token, setToken] = useState(initialToken || Cookies.get("token"));
  const [userId, setUserId] = useState(initialUserId || Cookies.get("userId"));
  const [chatState, setChatState] = useState("default");

  // If the token changed (i.e. the user logged in
  // or out), clear the Apollo store and refresh the
  // websocket connection.
  useEffect(() => {
    if (!token) client.clearStore();
    if (socket) refreshSocket(socket);
  }, [client, socket, token]);

  const setAuth = (data) => {
    if (data) {
      const { id, token } = data;
      Cookies.set("token", token);
      Cookies.set("userId", id);
      setToken(token);
      setUserId(id);
    } else {
      Cookies.remove("token");
      Cookies.remove("userId");
      setToken(null);
      setUserId(null);
    }
  };

  return (
    <AuthContext.Provider value={{ token, userId, setAuth }}>
      <ChatContext.Provider value={{ chatState, setChatState }}>
        {children}
      </ChatContext.Provider>
    </AuthContext.Provider>
  );
};

export default StateProvider;
