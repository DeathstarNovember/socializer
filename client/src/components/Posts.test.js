import React from "react";
import { render, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import tk from "timekeeper";
import { Subscriber } from "containers";
import { AuthContext } from "util/context";
import Posts, { GET_POSTS, POSTS_SUBSCRIPTION } from "./Posts";

jest.mock("containers/Subscriber", () =>
  jest.fn().mockImplementation(({ children }) => children),
);

describe("Posts", () => {
  beforeEach(() => {
    tk.freeze("2019-04-20");
  });

  afterEach(() => {
    tk.reset();
  });

  it("renders correctly when loading", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{}}>
          <MockedProvider mocks={[]} addTypename={false}>
            <Posts />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when loaded", async () => {
    const mocks = [
      {
        request: {
          query: GET_POSTS,
        },
        result: {
          data: {
            posts: [
              {
                id: 1,
                body: "Thoughts",
                insertedAt: "2019-04-18T00:00:00",
                user: {
                  id: 1,
                  name: "John Smith",
                  gravatarMd5: "abc",
                },
              },
            ],
          },
        },
      },
    ];
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{}}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Posts />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("Thoughts"));
    expect(container).toMatchSnapshot();
  });

  it("renders correctly after created post", async () => {
    Subscriber.mockImplementation((props) => {
      const { default: ActualSubscriber } = jest.requireActual(
        "containers/Subscriber",
      );
      return <ActualSubscriber {...props} />;
    });

    const mocks = [
      {
        request: {
          query: GET_POSTS,
        },
        result: {
          data: {
            posts: [
              {
                id: 1,
                body: "Thoughts",
                insertedAt: "2019-04-18T00:00:00",
                user: {
                  id: 1,
                  name: "John Smith",
                  gravatarMd5: "abc",
                },
              },
            ],
          },
        },
      },
      {
        request: {
          query: POSTS_SUBSCRIPTION,
        },
        result: {
          data: {
            postCreated: {
              id: 2,
              body: "Opinions",
              insertedAt: "2019-04-19T00:00:00",
              user: {
                id: 2,
                name: "Jane Thompson",
                gravatarMd5: "def",
              },
            },
          },
        },
      },
    ];
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{}}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Posts />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("Opinions"));
    expect(container).toMatchSnapshot();
  });
});
