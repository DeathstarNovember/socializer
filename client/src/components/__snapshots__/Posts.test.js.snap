// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts renders correctly after created post 1`] = `
<div>
  <h4>
    Feed
  </h4>
  <div
    class="mb-4 card"
  >
    <div
      class="card-body"
    >
      <div
        class="text-muted h7"
      >
        <a
          href="/login"
        >
          Log in
        </a>
         to submit a 
        post
        .
      </div>
    </div>
  </div>
  <div
    class="mb-4 card"
  >
    <div
      class="d-flex justify-content-between align-items-center card-header"
    >
      <div
        class="d-flex align-items-center"
      >
        <img
          alt="Gravatar for undefined"
          class="react-gravatar rounded-circle"
          height="30"
          src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
          srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
          width="30"
        />
        <div
          class="ml-2"
        >
          <div
            class="h6 mb-0"
          >
            Jane Thompson
          </div>
        </div>
      </div>
      <div
        class="text-muted h7"
      >
        <time
          datetime="2019-04-19T00:00:00.000Z"
          title="2019-04-19T00:00:00Z"
        >
          1 day ago
        </time>
      </div>
    </div>
    <div
      class="card-body"
    >
      <p
        class="card-text"
      >
        Opinions
      </p>
    </div>
    <div
      class="d-flex card-footer"
    >
      <a
        class="ml-auto"
        href="/posts/2"
      >
        Discussion ≫
      </a>
    </div>
  </div>
  <div
    class="mb-4 card"
  >
    <div
      class="d-flex justify-content-between align-items-center card-header"
    >
      <div
        class="d-flex align-items-center"
      >
        <img
          alt="Gravatar for undefined"
          class="react-gravatar rounded-circle"
          height="30"
          src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
          srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
          width="30"
        />
        <div
          class="ml-2"
        >
          <div
            class="h6 mb-0"
          >
            John Smith
          </div>
        </div>
      </div>
      <div
        class="text-muted h7"
      >
        <time
          datetime="2019-04-18T00:00:00.000Z"
          title="2019-04-18T00:00:00Z"
        >
          2 days ago
        </time>
      </div>
    </div>
    <div
      class="card-body"
    >
      <p
        class="card-text"
      >
        Thoughts
      </p>
    </div>
    <div
      class="d-flex card-footer"
    >
      <a
        class="ml-auto"
        href="/posts/1"
      >
        Discussion ≫
      </a>
    </div>
  </div>
</div>
`;

exports[`Posts renders correctly when loaded 1`] = `
<div>
  <h4>
    Feed
  </h4>
  <div
    class="mb-4 card"
  >
    <div
      class="card-body"
    >
      <div
        class="text-muted h7"
      >
        <a
          href="/login"
        >
          Log in
        </a>
         to submit a 
        post
        .
      </div>
    </div>
  </div>
  <div
    class="mb-4 card"
  >
    <div
      class="d-flex justify-content-between align-items-center card-header"
    >
      <div
        class="d-flex align-items-center"
      >
        <img
          alt="Gravatar for undefined"
          class="react-gravatar rounded-circle"
          height="30"
          src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
          srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
          width="30"
        />
        <div
          class="ml-2"
        >
          <div
            class="h6 mb-0"
          >
            John Smith
          </div>
        </div>
      </div>
      <div
        class="text-muted h7"
      >
        <time
          datetime="2019-04-18T00:00:00.000Z"
          title="2019-04-18T00:00:00Z"
        >
          2 days ago
        </time>
      </div>
    </div>
    <div
      class="card-body"
    >
      <p
        class="card-text"
      >
        Thoughts
      </p>
    </div>
    <div
      class="d-flex card-footer"
    >
      <a
        class="ml-auto"
        href="/posts/1"
      >
        Discussion ≫
      </a>
    </div>
  </div>
</div>
`;

exports[`Posts renders correctly when loading 1`] = `
<div>
  <h4>
    Feed
  </h4>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
